stablehlo/dialect/TypeInference.cpp

```diff
@@ -1484,8 +1482,57 @@ LogicalResult inferCholeskyOp(
         location, "minor dimensions of 'a' must have equal size, got shape ",
         aShape, ".");
   }
-  inferredReturnShapes.emplace_back(aRankedType.getShape(),
-                                    aRankedType.getElementType());
+
+  // bound infer rules for the last two dimensions of A:
+  //        Dim R-2 | Dim R-1 | inferred R-2 | inferred R-1
+  // case0: 3       | 3       | 3            | 3
+  // case1: ?       | 3       | 3            | 3
+  // case2: ?, A>=3 | 3       | 3            | 3
+  // case3: 3       | ?       | 3            | 3
+  // case4: 3       | ?, A>=3 | 3            | 3
+  auto resultBounds = encodingToBounds(aRankedType.getEncoding()).vec();
+  if (isStaticDimSize(lastDim) || isStaticDimSize(penultimateDim)) {
+    auto resultShape = aRankedType.getShape().vec();
+    auto staticSize = isStaticDimSize(lastDim) ? lastDim : penultimateDim;
+    resultShape[resultShape.size() - 1] = staticSize;
+    resultShape[resultShape.size() - 2] = staticSize;
+    if (resultBounds.empty()) {
+      inferredReturnTypes.push_back(
+          RankedTensorType::get(resultShape, aRankedType.getElementType()));
+    } else {
+      resultBounds[resultBounds.size() - 1] = ShapedType::kDynamic;
+      resultBounds[resultBounds.size() - 2] = ShapedType::kDynamic;
+      inferredReturnTypes.push_back(RankedTensorType::get(
+          resultShape, aRankedType.getElementType(),
+          boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
+    }
+    return success();
+  }
+  // case5: ?       | ?       | ?            | ?
+  if (resultBounds.empty()) {
+    inferredReturnTypes.push_back(RankedTensorType::get(
+        aRankedType.getShape(), aRankedType.getElementType(),
+        boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
+    return success();
+  }
+  // case6: ?, A    | ?       | ?, A         | ?, A
+  // case7: ?       | ?, A    | ?, A         | ?, A
+  // case8: ?, A    | ?, B    | ?, min(A,B)  | ?, min(A,B)
+  bool isLastDimStaticBound =
+      !isDynamicDimSize(resultBounds[resultBounds.size() - 1]);
+  bool isPenultimateDimStaticBound =
+      !isDynamicDimSize(resultBounds[resultBounds.size() - 2]);
+  int64_t staticBound = isLastDimStaticBound
+                            ? resultBounds[resultBounds.size() - 1]
+                            : resultBounds[resultBounds.size() - 2];
+  if (isLastDimStaticBound && isPenultimateDimStaticBound)
+    staticBound = std::min(resultBounds[resultBounds.size() - 1],
+                           resultBounds[resultBounds.size() - 2]);
+  resultBounds[resultBounds.size() - 1] = staticBound;
+  resultBounds[resultBounds.size() - 2] = staticBound;
+  inferredReturnTypes.push_back(RankedTensorType::get(
+      aRankedType.getShape(), aRankedType.getElementType(),
+      boundsToEncoding(aRankedType.getEncoding(), resultBounds)));
```

Comment:
Also, we'll want to document the rationale behind this change which is currently not obvious to me. E.g. what is the driving use case? At the moment, the PR description just links to the ticket which doesn't have a description provided.

Reply:
Can you clarify what change you are referring to? If you mean the rationale for changing the interface, I've reverted back to using InferTensorType as mentioned in the other comment.

Reply:
Ah okay. I realized that all this code is not needed, and is deleted in the recent commit as they are all related to type inference with dynamism not bounded dynamism. However for the last case8 which does have bounded dynamism, it is also conveniently handled by inferMostSpecificTypeComponents, so this PR ends up being a one line change.

Reply:
We cannot use any previous utility functions as Cholesky op deals with inference working with its dimensions not between types of multiple operands.

> At the moment, the PR description just links to the ticket which doesn't have a description provided.

I see that in a recent PR #799 that the discussions are happening in the issue not the PR. Since I've started some discussions already in the PR and not the issue, should I copy over and continue the discussion in the ticket? I'll provide early review in the issue itself for future shape functions.

Reply:
No strong preference on my side about the location of these discussions. If it's in the issue, that's fine. If it's in a PR, that's fine too. If it's split, that's great as well. As long as we make a note somewhere about the motivation behind the tricky changes :)